File: bindings/python/test/Library/Physics/Time/Instant.py
# coding=utf-8

################################################################################################################################################################

# @project        Library ▸ Physics
# @file           LibraryPhysicsPy/Time/Instant.py
# @author         Lucas Brémond <lucas@loftorbital.com>
# @license        Apache License 2.0

################################################################################################################################################################

import unittest

from Library.Core import Types

from LibraryPhysicsPy import Time

################################################################################################################################################################

Scale = Time.Scale
Instant = Time.Instant
DateTime = Time.DateTime

class TestLibraryPhysicsTimeInstant (unittest.TestCase):

    def setUp (self):

        self.instant = Instant.DateTime(DateTime(2018, 1, 1, 0, 0, 0, 0, 0, 0), Scale.UTC)

    def runTest (self):

        self.assertTrue(self.instant == self.instant)
        self.assertFalse(self.instant != self.instant)

        self.assertTrue(self.instant.isDefined())
        self.assertTrue(self.instant.isPostEpoch())

        self.assertEqual(self.instant.getDateTime(Scale.UTC), DateTime(2018, 1, 1, 0, 0, 0, 0, 0, 0))
        self.assertEqual(self.instant.getJulianDate(Scale.UTC), 2458119.5)
        self.assertEqual(self.instant.getModifiedJulianDate(Scale.UTC), 58119.0)
        self.assertEqual(self.instant.toString(Scale.UTC), "2018-01-01 00:00:00 [UTC]")

        self.assertFalse(Instant.Undefined().isDefined())
        self.assertTrue(Instant.Now().isDefined())
        self.assertTrue(Instant.J2000().isDefined())
        self.assertEqual(Instant.DateTime(DateTime(2018, 1, 1, 0, 0, 0, 0, 0, 0), Scale.UTC), self.instant)
        self.assertEqual(Instant.JulianDate(2458119.5, Scale.UTC), self.instant)
        self.assertEqual(Instant.ModifiedJulianDate(58119.0, Scale.UTC), self.instant)

suite = unittest.TestLoader().loadTestsFromTestCase(TestLibraryPhysicsTimeInstant)
unittest.TextTestRunner(verbosity=2).run(suite)

################################################################################################################################################################

File: bindings/python/src/LibraryPhysicsPy/Time/Instant.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

/// @project        Open Space Toolkit ▸ Physics
/// @file           LibraryPhysicsPy/Time/Instant.cpp
/// @author         Lucas Brémond <lucas@loftorbital.com>
/// @license        Apache License 2.0

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include <LibraryPhysicsPy/Utilities/IterableConverter.hpp>

#include <Library/Physics/Time/Instant.hpp>

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

inline void                     LibraryPhysicsPy_Time_Instant               ( )
{

    using namespace boost::python ;

    using ostk::core::types::String ;

    using ostk::physics::time::Scale ;
    using ostk::physics::time::Instant ;
    using ostk::physics::time::Duration ;

    scope in_Instant = class_<Instant>("Instant", no_init)

        .def(self == self)
        .def(self != self)

        .def(self < self)
        .def(self <= self)
        .def(self > self)
        .def(self >= self)

        .def(self + other<Duration>())
        .def(self - other<Duration>())
        .def(self - self)
        .def(self += other<Duration>())
        .def(self -= other<Duration>())

        .def(self_ns::str(self_ns::self))

        .def("__repr__", +[] (const Instant& anInstant) -> std::string { return anInstant.toString() ; })

        .def("isDefined", &Instant::isDefined)
        .def("isPostEpoch", &Instant::isPostEpoch)
        .def("isNear", &Instant::isNear)

        .def("getDateTime", &Instant::getDateTime)
        .def("getJulianDate", &Instant::getJulianDate)
        .def("getModifiedJulianDate", &Instant::getModifiedJulianDate)
        .def("toString", +[] (const Instant& anInstant) -> String { return anInstant.toString() ; })
        .def("toString", +[] (const Instant& anInstant, const Scale& aScale) -> String { return anInstant.toString(aScale) ; })

        .def("Undefined", &Instant::Undefined).staticmethod("Undefined")
        .def("Now", &Instant::Now).staticmethod("Now")
        .def("J2000", &Instant::J2000).staticmethod("J2000")
        .def("DateTime", &Instant::DateTime).staticmethod("DateTime")
        .def("JulianDate", &Instant::JulianDate).staticmethod("JulianDate")
        .def("ModifiedJulianDate", &Instant::ModifiedJulianDate).staticmethod("ModifiedJulianDate")

    ;

    using ostk::core::ctnr::Array ;

    IterableConverter()

        .from_python<Array<Instant>>()
        .to_python<Array<Instant>>()

    ;

}

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
